File: balanceddiet/balanceddiet.py



def getsum(descriptor,cans):
	s=0
	for i in xrange(len(cans)):
		if descriptor&1:
			s+=cans[i]
		descriptor>>=1
	return s


bits={}
b=1
for i in xrange(21):
	bits[b]=i
	b<<=1

lbits=sorted(bits)



def optimize(descriptor,index,cans,target,cache={}):
	if (descriptor,index) in cache:
		return cache[(descriptor,index)]
	if index==len(cans):
		cache[(descriptor,index)]=getsum(descriptor,cans)
		return cache[(descriptor,index)]
	m0=optimize(descriptor,index+1,cans,target,cache)
	m1=optimize(descriptor+(1<<index),index+1,cans,target,cache)
	m0,m1=sorted([m0,m1])
	cache[(descriptor,index)]= min([m0,m1],key=lambda x: abs(target-x))
	return cache[(descriptor,index)]





'''
cans=map(int,raw_input().split())

while cans.pop(0):
	s=sum(cans)
	cals=optimize(0,0,cans,s//2,{})
	x,y = sorted([s-cals,cals])
	print y,x
	cans=map(int,raw_input().split())
'''


cans=map(int,raw_input().split())

while cans.pop(0):
	s=sum(cans)
	combs=[0]*(1<<len(cans))
	for b in lbits:
		if b>=len(combs):
			break
		combs[b]=cans[bits[b]]
	for i in xrange(2,len(combs)):
		if combs[i]==0:
			rbit=i&-i
			combs[i]=combs[rbit]+combs[i-rbit]
	cals=min(combs,key=lambda v: abs(v-s//2))
	x,y = sorted([s-cals,cals])
	print y,x
	cans=map(int,raw_input().split())

File: balanceddiet/balanceddiet.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int bits[21];

int cache[1<<20]={-1};

void init_bits() {
	bits[0]=1;
	int i;
	for (i=1;i<21;i++) {
		bits[i]=bits[i-1]<<1;
	}
}

void init_cache(int *cans,int n) {
	int i;
	for (i=0;i<n;i++) {
		cache[bits[i]]=cans[i];
	}
}


void clear_cache() {
	int i;
	for (i=0;i<bits[20];i++) {
		cache[i]=-1;
	}
}


int getsum(int descriptor,int *cans) {
	if (cache[descriptor]>=0) {
		return cache[descriptor];
	}
	if (!descriptor) {
		cache[0]=0;
		return 0;
	}
	int rbit= (descriptor&-descriptor);
	cache[descriptor]=cache[rbit]+getsum(descriptor-rbit,cans);
	return cache[descriptor];
}


int optimize(int descriptor,int index,int * cans,
	int numcans,int target) {
	int m0,m1,temp;
	if (index==numcans) {
		return getsum(descriptor,cans);
	}
	m0=optimize(descriptor,index+1,
		cans,numcans,target);
	m1=optimize(descriptor+bits[index],index+1,
		cans,numcans,target);
	if (m0>m1) {
		temp=m0;
		m0=m1;
		m1=temp;
	}
	if (abs(target-m0)<=abs(target-m1)) {
		return m0;
	}
	return m1;
}


void read_can(int * cans, int n) {
	int i;
	for (i=0;i<n;i++) {
		scanf("%d",&cans[i]);
	}

}

int max(int x,int y) {
	return x>y? x:y;
}

int min(int x,int y) {
	return x<y? x:y;
}


int main(int argc, char const *argv[])
{
	init_bits();
	clear_cache();
	int n,cals,temp,s;
	int cans[20];
	scanf("%d",&n);
	while (n) {
		//printf("%d\n", n);
		read_can(cans,n);
		init_cache(cans,n);
		s=getsum(bits[n]-1,cans);
		cals=optimize(0,0,cans,n,s/2);
		printf("%d %d\n",max(cals,s-cals),min(cals,s-cals) );
		scanf("%d",&n);
		clear_cache();
	}
	return 0;
}





